File: Day263_2025/Q263_Leetcode.py
from collections import deque
import bisect
from typing import List

class Router:
    def __init__(self, memoryLimit: int):
        self.mpp = {}
        self.queue = deque()
        self.timestamps = {}
        self.st = {}
        self.maxSize = memoryLimit

    def addPacket(self, source: int, destination: int, timestamp: int) -> bool:
        packet = (source, destination, timestamp)
        if packet in self.mpp:
            return False

        if len(self.queue) == self.maxSize:
            res = self.queue.popleft()
            self.mpp.pop(res, None)
            temp = res[1]
            self.st[temp] = self.st.get(temp, 0) + 1

        self.queue.append(packet)
        self.mpp[packet] = 1
        if destination not in self.timestamps:
            self.timestamps[destination] = []
        self.timestamps[destination].append(timestamp)
        return True

    def forwardPacket(self) -> List[int]:
        if not self.queue:
            return []
        res = self.queue.popleft()
        self.mpp.pop(res, None)
        temp = res[1]
        self.st[temp] = self.st.get(temp, 0) + 1
        return list(res)

    def getCount(self, destination: int, startTime: int, endTime: int) -> int:
        if destination not in self.timestamps:
            return 0
        p = self.timestamps[destination]
        temp = self.st.get(destination, 0)

        right = bisect.bisect_left(p, startTime, lo=temp)
        left = bisect.bisect_right(p, endTime, lo=temp)

        return left - right

File: Day263_2025/Q263_Leetcode.java
import java.util.*;

class Q263_Leetcode {
    private Map<List<Integer>, Integer> mpp;
    private Queue<List<Integer>> queue;
    private Map<Integer, List<Integer>> timestamps;
    private Map<Integer, Integer> st;
    private int maxSize;

    public Q263_Leetcode(int memoryLimit) {
        mpp = new HashMap<>();
        queue = new LinkedList<>();
        timestamps = new HashMap<>();
        st = new HashMap<>();
        maxSize = memoryLimit;
    }

    public boolean addPacket(int source, int destination, int timestamp) {
        List<Integer> packet = Arrays.asList(source, destination, timestamp);
        if (mpp.containsKey(packet)) return false;

        if (queue.size() == maxSize) {
            List<Integer> res = queue.poll();
            mpp.remove(res);
            int temp = res.get(1);
            st.put(temp, st.getOrDefault(temp, 0) + 1);
        }

        queue.offer(packet);
        mpp.put(packet, 1);
        timestamps.putIfAbsent(destination, new ArrayList<>());
        timestamps.get(destination).add(timestamp);

        return true;
    }

    public int[] forwardPacket() {
        if (queue.isEmpty()) return new int[0];
        List<Integer> res = queue.poll();
        mpp.remove(res);
        int temp = res.get(1);
        st.put(temp, st.getOrDefault(temp, 0) + 1);
        return new int[]{res.get(0), res.get(1), res.get(2)};
    }

    public int getCount(int destination, int startTime, int endTime) {
        if (!timestamps.containsKey(destination)) return 0;
        List<Integer> p = timestamps.get(destination);
        int temp = st.getOrDefault(destination, 0);

        List<Integer> subList = p.subList(temp, p.size());
        int right = Collections.binarySearch(subList, startTime);
        if (right < 0) right = -(right + 1);

        int left = Collections.binarySearch(subList, endTime);
        if (left < 0) left = -(left + 1);
        else left += 1;

        return left - right;
    }
}
